src/test/java/org/owasp/esapi/reference/EncoderTest.java

```diff
@@ -1050,4 +1051,176 @@ public void testGetDefaultCanonicalizationCodecs() {
 
         assertEquals( true, caughtExpected );   // Verify it threw an IntrusionException
     }
+
+    /**
+     * Test of encodeForJSON method, of class org.owasp.esapi.Encoder.
+     */
+    public void testEncodeForJSON() {
+        System.out.println("encodeForJSON");
+        Encoder instance = ESAPI.encoder();
+
+        // Empty strings
+        assertEquals( null, instance.encodeForJSON(null) );
+        assertEquals( "", instance.encodeForJSON("") );
+        assertEquals( " ", instance.encodeForJSON(" ") );
+
+        // Walk a message without escaped characters
+        String message = "Now is the time for all good men to come to the aide of their country.";
+        for( int i=1; i<message.length(); ++i ) {
+            final String substring = message.substring(0, i);
+            assertEquals( substring, instance.encodeForJSON(substring) );
+        }
+
+        // Two-character sequence escape representations of some
+        // popular characters
+        assertEquals( "\\b", instance.encodeForJSON("\b") );
+        assertEquals( "\\f", instance.encodeForJSON("\f") );
+        assertEquals( "\\r", instance.encodeForJSON("\r") );
+        assertEquals( "\\n", instance.encodeForJSON("\n") );
+        assertEquals( "\\t", instance.encodeForJSON("\t") );
+        assertEquals( "\\\"", instance.encodeForJSON("\"") );
+        assertEquals( "\\/",  instance.encodeForJSON("/" ) );
+        assertEquals( "\\\\", instance.encodeForJSON("\\") );
+
+        // All Unicode characters may be placed within the quotation marks,
+        // except for the characters that MUST be escaped: quotation mark,
+        // reverse solidus, and the control characters (U+0000 through U+001F).
+        for( int i=0; i<=0x1f; ++i ) {
```

Comment:
Very concise... I really like this.  First thing I was checking for was if all characters were being tested.  

We MAY run into downstream issues from the fact that the choice of AbstractCharacterCodec instead of AbstractIntegerCodec will cause UTF-8 encoded characters to be destroyed.  

What happens when the character is 😂, otherwise known as U+1F602?  

Could you add a unit test that covers this case at least, and that we we have a documented test case in code at least that records a known failure? (I suspect that 😂 will be downcast into gibberish)

Reply:
> What happens when the character is joy, otherwise known as U+1F602?

I checked-in a test case for this. The codec can round-trip it through the encoder and decoder.

I'm still not convinced things are correct since a Java string is UTF-16, but this codec should operate on UTF-8.